files/default/neutron-ha-tool.py

```diff
@@ -576,6 +600,195 @@ def l3_agent_evacuate(qclient, agent_host, agent_picker, router_filter,
+            agent_to_evacuate = agent
+            break
+
+    if not agent_to_evacuate:
+        LOG.error("Could not locate agent to evacuate; aborting!")
+        return 0
+
+    target_list = list_alive_dhcp_agents_except(agent_list,
+                                                exclude_agent=agent_to_evacuate)
+
+    if len(target_list) < 1:
+        LOG.error("There are no DHCP agents alive to migrate routers onto")
+        return 0
+
+    (migrations, errors) = \
+        migrate_dhcp_networks_from_agent(qclient, agent_to_evacuate,
+                                      target_list, agent_picker,
+                                      noop, wait_for_network, fail_fast)
+    LOG.info("%d networks %s evacuated from DHCP agent %s", migrations,
+             "would have been" if noop else "were", agent_host)
+    if errors > 0:
+        LOG.error("%d errors encountered during evacuation", errors)
+
+    return errors
+
+
+def migrate_dhcp_networks_from_agent(qclient, agent, targets, agent_picker,
+                                     noop, wait_for_network, fail_fast):
+    LOG.info("Querying agent_id=%s for networks to migrate away", agent['id'])
+
+    networks_on_agent = qclient.list_networks_on_dhcp_agent(agent['id'])['networks']
+
+    agent_picker.set_agents(targets)
+
+    migrations = 0
+    errors = 0
+    for network in networks_on_agent:
+        target = agent_picker.pick()
+
+        if noop:
+            LOG.info("Would try to migrate network=%s from agent=%s "
+                 "to agent=%s", network['id'], agent['id'], target['id'])
+            migrations += 1
+            continue
+
+        try:
+            migrate_dhcp_network(qclient, network, agent, target, wait_for_network)
+            migrations += 1
+        except:
+            LOG.exception("Failed to migrate network_id=%s from agent=%s to "
+                          "agent=%s", network['id'], agent['id'], target['id'])
+            errors += 1
+            if fail_fast:
+                break
+
+    return (migrations, errors)
+
+
+def migrate_dhcp_network(qclient, network, agent, target,
+                         wait_for_network=True):
+    """
+    Returns nothing, and raises exceptions on errors.
+
+    :param qclient: A neutronclient
+    :param network: The network to migrate
+    :param agent: The DHCP agent to migrate/remove from
+    :param target: The DHCP agent to migrate to
+    """
+
+    LOG.info("Migrating network=%s from agent=%s to agent=%s",
+             network['id'], agent['id'], target['id'])
+
+    # Prevent SIGTERM only when modifying data through the API, and allow it
+    # when ensuring the API operations completed successfully
+    with term_disabled():
+
+        # Remove the network from the original agent
+        qclient.remove_network_from_dhcp_agent(agent['id'], network['id'])
+        LOG.debug("Removed network from agent=%s" % agent['id'])
+
+        # ensure it is removed
```

Comment:
So `remove_network_from_dhcp_agent` does not return anything reasonable so that we have to check the result by another call?

Reply:
I followed the pattern from routers migration here. As far as I can see in `neutronclient` code, these functions are just thin wrappers around API calls and return values represent the HTTP return codes, so I guess it can only return something like "the request was accepted" but not "the network was actually removed from agent".